firmware/tracing_firmware/main.rs

```diff
@@ -0,0 +1,178 @@
+#![no_std]
+#![no_main]
+
+use core::arch::{asm, global_asm};
+
+use miralis_abi::{failure, log, setup_binary, success};
+
+setup_binary!(main);
+
+fn enable_mcycle_in_smode() {
+    unsafe {
+        let mcounteren: u32;
+        asm!("csrr {}, mcounteren", out(reg) mcounteren);
+        asm!("csrw mcounteren, {}", in(reg) mcounteren | 1);
+    }
+}
+
+fn main() -> ! {
+    unsafe {
+        asm!("la sp, 0x80700000");
+    }
+
+    enable_mcycle_in_smode();
+
+    let os: usize = operating_system as usize;
+    let trap: usize = _empty_handler as usize;
+    let mpp = 0b1 << 11; // MPP = S-mode
+
+    unsafe {
+        asm!(
+        "li t4, 0xfffffffff",
+        "csrw pmpcfg0, 0xf",   // XRW TOR
+        "csrw pmpaddr0, t4",   // All memory
+        "auipc t4, 0",
+        "addi t4, t4, 24",
+        "csrw mtvec, {mtvec}", // Write mtvec with trap handler
+        "csrw mstatus, {mpp}", // Write MPP of mstatus to S-mode
+        "csrw mepc, {os}",     // Write MEPC
+        "mret",                // Jump to OS
+
+
+        os = in(reg) os,
+        mtvec = in(reg) trap,
+        mpp = in(reg) mpp,
+        out("t4") _,
+        );
+    }
+
+    success();
```

Comment:
If the OS is supposed to exit I would suggest to use `failure()` here, as it means we failed to jump into the OS.